Blobs mirroring (Look up blobs using HTTP in parallel with DHT lookup)

shyba opened the pull request:
For https://github.com/lbryio/lbry/issues/1212

Original title is "Look up blobs using HTTP in parallel with DHT lookup" and original idea was to do it as soon as possible. I tried my best to do it in a way that can be called anywhere in the codebase (as long as the caller has a way to get a blob_manager) and implemented in a middle ground between "when it fails" and "right on start". We can discuss where it's best to move it and I also tried to make it so it's easy to completely remove from our code (nobody really depends on it, they just ask help and ignore the result without even waiting for the deferred to fire).

One should notice that there is an unused method called `get_all_missing_blobs`. I added for the sake of having it so we can call during `blob_list --needed`, for instance, to fix all missing ones, however the reviewers are free to ask its removal if after discussing we find no usage for that.

kauffj commented on lbrynet/daemon/Downloader.py:
```diff
@@ -84,6 +85,8 @@ def check_status(self):
         else:
             log.debug("Waiting for stream descriptor (%i seconds)", self.timeout_counter)
+            if self.mirror:
+                self.mirror.get_blob(self.sd_hash)
```
Calling `check_status` can trigger a blob download? Seems like a side-effect.

kauffj commented on lbrynet/core/client/BlobRequester.py:
```diff
@@ -142,6 +144,8 @@ def choose_best_peers(peers):
             without_bad_peers = [p for p in peers if not p in bad_peers]
             without_maxed_out_peers = [
                 p for p in without_bad_peers if p not in self._maxed_out_peers]
+            if not without_bad_peers:
```
Calling `choose_best_peers` can trigger a blob download? Seems like a side-effect.

lyoshenka commented on lbrynet/core/BlobMirror.py:
```diff
@@ -0,0 +1,57 @@
+
+class BlobMirror(object):
+    def __init__(self, blob_manager, servers=None, client=None):
+        self.servers = list(servers or conf.settings['blob_server_domains'])
```
id prefer to require that servers be passed in. then we can avoid a dependency on conf, which is already all over our code

shyba commented:
Changed to concurrently download, added tests and tried to better integrate to the existing organization so it become less of a hack, but still not much intrusive as a refactor on p2p is expected to happen soon.

eukreign commented on lbrynet/core/HTTPBlobDownloader.py:
```diff
@@ -0,0 +1,81 @@
+
+    def start(self):
+        if not self.running and self.blob_hashes and self.servers:
+            return self.looping_call.start(self.interval, now=True)
```
Shouldn't this return `defer.succeed(True)` in the implicit `else` case? Otherwise it requires the caller to `defer.maybeDeferred()` your `start()` method. I think it's good etiquette to keep an `async` method `async` in all circumstances.

shyba replied:
Fixed, thanks. (the comment isn't gone because it's the other line)

eukreign commented on lbrynet/core/HTTPBlobDownloader.py:
```diff
@@ -0,0 +1,81 @@
+    def _download_next_blob_hash_for_file(self):
+        for blob_hash in self.blob_hashes:
+            blob = yield self.blob_manager.get_blob(blob_hash)
+            if not blob.get_is_verified():
```
API should just be an `@property` with name `is_verified` instead of `get_is_verified()` method call.

eukreign commented on lbrynet/core/HTTPBlobDownloader.py:
```diff
@@ -0,0 +1,81 @@
+            blob = yield self.blob_manager.get_blob(blob_hash)
+            if not blob.get_is_verified():
+                self.download_blob(blob)
+                defer.returnValue(None)
```
You can just put a regular `return` statement, it will have the same effect as `defer.returnValue(None)`.

eukreign commented on lbrynet/core/HTTPBlobDownloader.py:
```diff
@@ -0,0 +1,81 @@
+    def download_blob(self, blob):
+        d = self._download_blob(blob)
+        d.addCallback(self._on_completed_blob)
+        d.addErrback(self._on_failed)
```
I think we're moving towards using `@defer.inlineCallbacks` more because it's easier to read / reason about and will make a some-day port to asyncio easier. And in this particular case it will have the added advantage of reducing the class surface area by moving the two methods below directly inline into this method.

eukreign commented on lbrynet/core/StreamDescriptor.py:
```diff
@@ -445,7 +445,10 @@ def download_sd_blob(session, blob_hash, payment_rate_manager, timeout=None):
                                           session.wallet,
                                           timeout)
+    mirror = HTTPBlobDownloader(session.blob_manager, [blob_hash], session.download_mirrors)
+    mirror.start()
```
Does this need to be `yield`ed?

shyba replied:
No, the idea is that it doesn't care about the result of `start` at all. Yielding would also make it wait, but we agreed to leave it concurrently.

eukreign commented on lbrynet/file_manager/EncryptedFileDownloader.py:
```diff
@@ -123,6 +127,10 @@ def _start(self):
+        if self.mirror:
+            blobs = yield self.storage.get_blobs_for_stream(self.stream_hash)
+            self.mirror.blob_hashes = [b.blob_hash for b in blobs if b.blob_hash is not None]
+            self.mirror.start()
```
`yield`?

shyba commented:
Thanks for the review, @eukreign! I think I fixed them.

lyoshenka commented on lbrynet/core/HTTPBlobDownloader.py:
```diff
@@ -0,0 +1,79 @@
+            if blob_hash:
+                log.debug('Mirror completed download for %s', blob_hash)
+            self.failures = 0
+        except Exception as exception:
```
all exceptions? shouldnt we be more specific?

eukreign replied:
We don't have a way to know what are all of the possible exceptions that could be thrown. I think the goal here is to retry downloading no matter what and log the specific exception.

eukreign replied:
`exception` variable still holds the actual specific exception, the error message will include the specific exception in the output. This is basically a catch all to make sure we keep jugging along no matter what might have failed.

shyba replied:
Yes, in general, but I believe not in this case. This loop doesn't raise to callers, so if the exception isn't handled and logged here it won't be anywhere and could cause cryptic "unhandled error" messages. IMO, but if you disagree we can be specific.

tzarebczan commented:
My only complaint would be to suppress or fix this error: 
```
2018-06-20 23:47:10,417 INFO     lbrynet.core.client.BlobRequester:486: Requesting blob 46ab794e4e5d897fcfe61d239a24ba77325ececb206ef4302360b3a4dffec4983495ffa41022d1a80b9bec88c4332144 from 54.209.131.62:3333
2018-06-20 23:47:10,430 ERROR    lbrynet.blob.writer:30: writer has already been closed
Traceback (most recent call last):
  File "site-packages\lbrynet\daemon\DaemonControl.py", line 78, in start
  File "site-packages\jsonrpc\proxy.py", line 176, in __call__
  File "site-packages\jsonrpc\proxy.py", line 169, in _post
  File "urllib2.py", line 429, in open
  File "urllib2.py", line 447, in _open
  File "urllib2.py", line 407, in _call_chain
  File "urllib2.py", line 1228, in http_open
  File "urllib2.py", line 1198, in do_open
URLError: <urlopen error [Errno 10061] No connection could be made because the target machine actively refused it>
2018-06-20 23:47:10,444 ERROR    lbrynet.core.HTTPBlobDownloader:54: Mirror failed downloading:
2018-06-20 23:47:10,561 INFO     lbrynet.core.client.BlobRequester:562: Blob 46ab794e4e5d897f has been successfully downloaded from 54.209.131.62:3333
2018-06-20 23:47:10,563 INFO     lbrynet.core.client.BlobRequester:486: Requesting blob 50641b12999c8642ed97f0587d36945d8a611a143326f73dbde57d529a862218ac1b37c
```

tzarebczan commented:
@shyba see above - do we want to do any about the above error?

tzarebczan commented:
Also, for downloads (and their associated blobs) that are still incomplete, I get a "missing a blob" error - see below for my daemon startup. Maybe we should suppress these? 

```
2018-07-10 11:41:04,085 INFO     lbrynet.daemon.Daemon:263: Started lbrynet-daemon
2018-07-10 11:41:04,596 INFO     lbrynet.reflector.client.client:77: No more completed blobs for 3ce356d96400be5b to reflect, 4 are still needed
2018-07-10 11:41:04,598 INFO     lbrynet.reflector.client.client:77: No more completed blobs for bafc8dcaecfb8555 to reflect, 3 are still needed
2018-07-10 11:41:04,611 INFO     lbrynet.reflector.client.client:77: No more completed blobs for de6b8f8f7c981715 to reflect, 8 are still needed
2018-07-10 11:41:04,686 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: cb1de038e53d29fb42608627cbcdd1a68cd67a895937611c8064fa09854cfc8690a3f0b316466581bcfdcdc145cceb8b
2018-07-10 11:41:04,700 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: c1e314936e1da506c5b4ab099f379939b49e1bc5677efb373dc94b1b7333db50483aa1b7e597441d036db2fd6b82bfbc
2018-07-10 11:41:05,194 INFO     lbrynet.reflector.client.client:77: No more completed blobs for 1e8bda659de87088 to reflect, 173 are still needed
2018-07-10 11:41:06,035 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: fddf102dc6ff19ddd1fbbd2ad042a3a778205c0ad9c3e5b628c505c5c98bc155be93a1a9efbbe0557dcec302eeb92f1d
2018-07-10 11:41:06,075 INFO     lbrynet.dht.node:247: Attempting to join the DHT network, 9 contacts known so far
2018-07-10 11:41:06,121 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 46e08f959936fc5e7cc405903e9218f200efe2b11138643df5392bc6dba4162ac19438eeb5d4f58287f8844d5a7fc0bd
2018-07-10 11:41:06,417 INFO     lbrynet.reflector.client.client:77: No more completed blobs for 16fd13e23db4661a to reflect, 859 are still needed
2018-07-10 11:41:06,461 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 968a724c1d79cee553bc14f28ae5dccf8dbf758e8327e79d04450d601e65e407711084321e47d2dc97b61c3412d23746
2018-07-10 11:41:06,618 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 2cf2affaaa948056e45ef6d68fc5a383f33c1c586615220aa50f25c7ec4a7130010303147d68d557daf4fa6d86d950a6
2018-07-10 11:41:07,457 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e9a0077f19afc961347f9bc77a9d79f4aa895e4ec26390827d581052ddf620923af8deea3424d1760078a1a49f0792a9
2018-07-10 11:41:07,615 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 70de0782ef1189d9c455e8a85bd68333ccbb1332126bad1fb87ccfd79e148924e34e629e0662a457fc93b7c9725cb7ce
2018-07-10 11:41:08,450 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: ac015a1216fd5958bd856fe1df5870a53463ce89fb3708d8157d1d6171dfb9ac5230ca7cfd86af39a6f32041465c966d
2018-07-10 11:41:08,618 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e4b9a54cfc1459246b85fc92103bccc29ac423cb8ee2ec5da0c44a127b586329aee10128dc64d6f5652b6c2f643ede73
2018-07-10 11:41:09,470 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e2bf2a667a4adde82568bd374a9a74cf5352180cff053cb6aecf889f4ffcc5ccb60dc861d1bc05855e16d3e7bdd1c2ef
2018-07-10 11:41:09,621 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 6029b918319a2fc9e8f1e935db8089ad8b75f533e5a87a1cd85ebc12bd19a41037c2dc929dfc1d87366d18945a1a6cf3
2018-07-10 11:41:10,486 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 4698198ef82fd52a0b79e6ead4d53ee9b4dea8e12e6e38046f02abd53ec4087b64ac6608c53975fea13b267565ca3aec
2018-07-10 11:41:10,653 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 25ad07c03f0528b2bc657bb651d6242a32626b8948678567bda861d7eff69b97abb5b610c6d4a20a0982d80aeecffefb
2018-07-10 11:41:11,457 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 566a5038d3912b6d45ef3bf0683c8d33e8118c9972af44ef34f84ba5a294193a751ef85daa74e044875dce1249ca48b4
2018-07-10 11:41:11,612 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 438f80c1c434b68c721d840658a7783fe323350ff3b983d9366f6a3a48b44a84ae335c422b72950cbcfa4e8a28cc7b59
2018-07-10 11:41:12,483 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 9399d6a84f3c4447b43e2415cf024c679d50913863cb72b653ad3b65b3a35133883926605f4d9e8d1c787e4cc3b11899
2018-07-10 11:41:12,983 WARNING  lbrynet.core.client.StreamProgressManager:158: Error outputting blob 34694830635750692d877580a1aabfcf757e71a26c65e62e907bc05ddbaaa404fe968062e5133191d32d3d0f1973b61c: Invalid padding bytes.
2018-07-10 11:41:13,226 WARNING  lbrynet.core.client.StreamProgressManager:158: Error outputting blob 2f87112c765f32baf8ec99b8379e850387a4596df832a7897377f998cf6f98f80413460a6a46766ba0081cf1a547cde6: Invalid padding bytes.
2018-07-10 11:41:13,444 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 60c0314123e4966b115ae91784527359e43c7a163188d4c7294aecf44f101cabd7b5ff7d3ecd3db31fcc2f34f4e4dc6a
2018-07-10 11:41:13,974 WARNING  lbrynet.core.client.StreamProgressManager:158: Error outputting blob bfe5f644d0af2c4a4bc3ec04a9b5260e2625a64b608bfef2819df6e81268567e0c2b35f52e710088db9d211ef8094328: Invalid padding bytes.
2018-07-10 11:41:14,454 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 8dcf6e9a1cc8775b4b1dcb0e14cf6fa7ffe061a36f3c86d7ffccda018aab479f9a997c46e5f226a676b7a5d917cafdbc
2018-07-10 11:41:14,789 INFO     lbrynet.core.Session:232: Joined the dht
2018-07-10 11:41:15,463 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 47b3c28b897f3114242edcbb5889c66183162cac6e309b5bbeda9ef81b72cac11ebb00efa0b2b06f7b8f0872ca983832
2018-07-10 11:41:16,457 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 1a72887e4d498bd587b040346299b395bf6d3060f08313a2836e837bbe127cc4325da992d42eefc7621cdd93d8d88e3b
2018-07-10 11:41:17,463 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 4b5be7ba2634f3b0bbd9efdd60aae068db123bfec3e7866ee9320f180fd7cb422ad3a8a65f6e54752ec17afcc63c5149
2018-07-10 11:41:18,444 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 2d6c86b05c64669556f3e681ed967082c52e4a225b76ed86e7a9bdfb09c9cc7e39d3ed43ad9238416611c4c767b6a88f
2018-07-10 11:41:19,437 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b3a7307d4d38036cd98731914b9234b471132f28d44c47d49bfcf0d333eaab1724aaabd7752b138df2ec529f19c71381
2018-07-10 11:41:20,444 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: fb483d7374f7afe5e35d2f7bb83f299930db59dc6b1e62bc66325fc894fde0c52b74dea26787d7dcf94d2c01dc1ee8d6
2018-07-10 11:41:21,454 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 817a183813747a487f2e07b5e4ed9ccdf8e42c2bbba4f448cf84835b3d38ab789b48b00ef3207d9120646c1ab005cf44
2018-07-10 11:41:21,704 WARNING  lbrynet.core.client.StreamProgressManager:158: Error outputting blob c025ca07aa5b9a68a1bc878de04791f6e2bd9c6436452df03940e51e51a11834220a5594bfc8fdb4011f96be1f702273: Invalid padding bytes.
2018-07-10 11:41:22,434 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 7f562c2f611f8f5bf8e8d50bcbb0685a6e0a7b67c6f68fa80919efa817f8fe2f348ea3678d6440b5835ac27ae1fee3f3
2018-07-10 11:41:23,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: eb0005be0aaf2ceab5dde91f89dd62ab9ced4b221f22e49a93dd0236152fd5cfe6627a4050c8caa9b3d3a69f8244cc1c
2018-07-10 11:41:24,502 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: a9a20be0e07198feb2a811d92c32ede8a77331ca99631140efa8d2ec38bb4cfe0345bc06dbe2d276e50d86fe273bd200
2018-07-10 11:41:25,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f0f65080401fc7fbac1954a235aee83a9229ac937c5eeb78e15901a3346640d38cff418530fc4f09e0a480915f097f9c
2018-07-10 11:41:26,476 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: debc7d85f29c78a9dc242e7b5a71b03d5e85818523956d2ac928fb51af42ad572d6be1b7493c9f0f6522a21c1fcff619
2018-07-10 11:41:27,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 37edb4cdd1853e1b222d946171d4a2673c19d46d541e5c27e1d6b8130e3f843375a8d5d9653e1eaffc47a27c2432fc01
2018-07-10 11:41:28,423 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f8c47a3444d3830a70ff6b090ca627987e2f1ca3098b4a8f5305f9b6c39053c56e9027f99dcfa086ed08e2afaed5a1dc
2018-07-10 11:41:29,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 335271ee1e9d5b1bdc37c9f5f1508108bb823311b7572e63c6479ff9c85529c57298c9e257ec5416fbb602025c833656
2018-07-10 11:41:30,431 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: cdb535fea1fd496298265d5e43b36a35b96d1d73449d8906b788ef5a36b2d25a94c45edcf5524d20d2cffc6a58c14eff
2018-07-10 11:41:31,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 396a33075cdfd070ab4bd1c55b82882ecc1cbd48703dd45e5107dcb437baf1902f925e87fa95e93c1d434a383d4aafe2
2018-07-10 11:41:32,505 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 96ae01278cb29e52b8a0f49666542737ac17eb7337d303ccedb265cd52dfd416d4c7f38591896aa3292d615f0e59a8f4
2018-07-10 11:41:33,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 4a081ad0ed67bbaf2e09ac3819271755920b58afe4fb3ce1ad31030fd2bb79815e41ff29283ee3ac69d8858b943bb980
2018-07-10 11:41:34,467 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 77f875da4d9a5301da44bf81f9c88054e41d764dd2b81939ab7d039c5df2a6a1728e09afb1e6e75dfb6c89d30fbd07a7
2018-07-10 11:41:35,430 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 07e265d2b689ec69d6ddde53ebf5e6696923d3ad1ba132a0e100ec0ae6581151251a98cf114e344470686bb340840b9c
2018-07-10 11:41:36,467 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: cc49149a5ef0246359fcd9ccc3e419617f96a5ec87b5dd3ed2e077a625799cd27c1cb40b3bad581a11f887f7d0a333f4
2018-07-10 11:41:37,423 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 6138949020ffdb7fdc871dce0e3290fc82230c39702a45fe24dd3f2f984b60bc5b22b7b78af3cffa58528cbc1b02eda8
2018-07-10 11:41:38,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: cb2baf9a0a821db63944e1347d378493af194dd40c8481bf8862556d72cb0c53a8f1dcef2a740f62fd7235ec735b9a7c
2018-07-10 11:41:39,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 5a9e91e920f5c4716cd97aefcfddbabf753359550aab643d38cd7dd746c529e3a6d5630d1323291793e0ba805544d928
2018-07-10 11:41:40,506 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 667e1266ee89e724dafb8c2ad7a51e6c614e09df3803a607bb04baeb1bfd19bdc773ea40ba6cc5218c2c3326b7bd3d0f
2018-07-10 11:41:41,421 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 9a372aef0b0b95829ba1ba3e729765291bde81ea3121e69fdcfc70dece16afcfbf408a5b242117c220bb3792b0f21764
2018-07-10 11:41:42,421 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 1f875b023bf8028b7f044b281d91ac0fbc9b721287bb91ab5fc5814f218e28dfdbd748d58dfb2214f5aa460f68426796
2018-07-10 11:41:43,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b226cd1f9ffbace3f4428d873249c7d12cfec48df7d9d4da5895b9273ae7a4d65abb1191a660c2013cc0d02e8560dd67
2018-07-10 11:41:44,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 9b32b0729acaf5a4d8b2666b656a10dd8f298e03755eb6f851d138c5b75b33356b5873d306a587d7f037d16849bc4d04
2018-07-10 11:41:45,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 41747039aea8b229877cc4193bb0271430422db733ea98cb0dcde8f4230e3362563dd33101da93db42df8c28fbf1e7f4
2018-07-10 11:41:46,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: c983c7baf3b15e1f0b91329d1733d9351f7d76b1b2254c4452774c2f14194663bd94e8527eff582c3b06ef7aca726f41
2018-07-10 11:41:47,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 86ede2085baa37449cd5a820c791d755814e58c88802c9a10a0fd99100d1605951b693afba15f67abadd56e2bb87cad2
2018-07-10 11:41:48,503 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 19ae2a9a9d785d55e36d049a28903e0ea183d473b32798dde9f6a1d498877546a641e38a782c05313d9a0d23b1596968
2018-07-10 11:41:50,210 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e6753ec3715fa615365e908c5cacc8f40c59bcbadc517a1d6befde1952bae1c5a42b8b0157136e3c92e4bae2d0baffbf
2018-07-10 11:41:50,471 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 1c49a5edba75a5732a7c6aa423c5f89f86be54f8cac8fae33a9104559e4f721e16b03a3f95ba44ef8f8b20d7565b0759
2018-07-10 11:41:51,434 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: d32b8633c4bbcc4c9a53e38d21ef514937512df593992248a78cdab363f9c6e8fbf53a638f53619f0a532d8744ecb83a
2018-07-10 11:41:52,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 5cb15a8b0de496c27e1439c6bc28ee483a241ed43d9f05876eb0532b7d1f476b292a0cd7504cf4d6603212521c79682c
2018-07-10 11:41:53,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e7a85a8470f001803d743fefc7eeef429b21666ff9d7d80e98e5c1ca79ef0908165f5251c50ef66a6f027e773cfc806c
2018-07-10 11:41:54,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 4ffcc6226ab54a208325cf86da6c0a62a8070938b1d035c5ceec0ad931394b20f34a89a274b16fb6660b267c4c336a95
2018-07-10 11:41:55,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 0333ca7bef4e3f485320a12dd74cdbb2fed1e7945830ae2a1f761e6beb0dbad372570384ea1946191e89a17ab560b0a7
2018-07-10 11:41:56,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: fe1210cd54518de2839d1fa80c56f1b019b9393033b83f8d1d65fcb9f41e5bfb5a642a2832cae7dcfbe73b1235447283
2018-07-10 11:41:57,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b05ce153ac31ba107238d39292aa58c68661c2c976b0bd8fc2e68443dd4f107650a8ea9698e4c8ad58a6037077b1e58c
2018-07-10 11:41:58,506 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 7fb4511cbb88e3b87ae7803da9edfb77eaf4509683fda9be22fa0abd80858a8e1bfa97152f66d9d501db82d24bc25261
2018-07-10 11:41:59,438 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 97909345d7856bc1826c7c9f43b8c305d3eb2ee63cf313f2b87fc29570a076cc78dc7c4a31aa3382cc2234f06c6f01dc
2018-07-10 11:42:00,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 177bac0a91e632517b110ad9d2936eda9d0c2e6438f38648a34f11e8d801e745eb7cbb03daa6f2657bef99d81963d2b1
2018-07-10 11:42:01,433 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 3e4307033879cebacab05baa4e1ced2ecd74e252acdd3de09eeb6d1382ce2464e568b0f705a7c4dd14c61787966aeb43
2018-07-10 11:42:02,423 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 45ef91c4ed16216401f8cbb193491e8a88e8731d7fdb0f65e15bebbc2aa2431b03fe73417e71c5d28cca9510c88631c3
2018-07-10 11:42:03,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 5b7ec3446dd3b11c08b2b7cf3ae5cd95db4caecfa7296219fdc14303da56163873e9d0cb08b32707936189cb2fc997cd
2018-07-10 11:42:04,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 47ec516aa0e64b2e32d0f38a740d34082d88cc45d6455a4e09179077d3286bd28f689c8beafd98f121125aab63f98fa0
2018-07-10 11:42:05,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b6e5485e31ff829bdb2e2db5479789226904ed19c1865fa1b4425f0504cb3a5292288f79e8132944e7b59c40313a5046
2018-07-10 11:42:06,502 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 1fe2d09d95bee3e13fb44165083de7012d121bdb4b010fd9e44146397f1cfe294517bb9e9496c29f634d3a33d93623a2
2018-07-10 11:42:07,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 13bc3e42a63eb30e656637f9d1b02895af30f8856463178263baee3e444859522021c5bb6bfc30bce02788387fbde59e
2018-07-10 11:42:08,433 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 24498beb4248dea0cb8066796612f579a1b548f37e76fc2b2ad0cfa22659e321e13972df3e14ac92125c59df08d2d299
2018-07-10 11:42:09,467 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f0f66376454b205b729148418fcf80e70c905a6ccfd6f85da1c7df7cdfd8610817c42be3f34c6bb7c60f3a58994b71d4
2018-07-10 11:42:10,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: bbf282d4036be251df5236e93f55ad631e60b98cb08148cfcc1c30e94ba4763cf76941543aaa95173f83b8ab43a33c4a
2018-07-10 11:42:11,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 8336d9b7d0bce02fed9e088d56691c19346552e27dcd46e9bd4b0380d37435cc60fe2da3380264b1c274573660d6b35d
2018-07-10 11:42:12,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b2418434a1e395b39c95cc5a4e65a3dadc84eb9a13bf7dac8f01e8504b1f8279dec2cf3daf43c6f2d78643e64df360ad
2018-07-10 11:42:13,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b8ba2c5f54173fdaafd7bcff61e10429067091bb6f39e3e452b10240401bf8b4402f8625e979036825aae66cd196bbe1
2018-07-10 11:42:14,505 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: af6234dd35787e279550b9f543a1f968141fd9d95c5ffce70ca9bff3c3883b0d72dc06bd78e737249fc40b5674d0adbb
2018-07-10 11:42:15,483 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 48c22e7dbd020ac732704ef81fd15e742b13b8200500d79704dab39b41396191f9ecfa56ef274a4a6ea372034ff750d5
2018-07-10 11:42:16,430 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: c25856c975bc3a32d2efc7db0053be3a4ddf9ec38f3b1c2d6abd140d758808887331a34cea53e4b559594db5c5210d0c
2018-07-10 11:42:17,430 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 83f59e7b8bdee4da61753b9aed6f4de6ae782d7fc5d3f61401be6aa1ad0ccf7964a637a05df6214ab487ea5f313cca6f
2018-07-10 11:42:18,423 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 42c0974587e15c5e9d953da3cd31bfd4629e732de86b98f0d4ffaf244dcb7379b65dacf284cf31106e765d719a987bdd
2018-07-10 11:42:19,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f9838a07c42ac9004f7f394b86ba80a5d9731ebf3c0abffe0a4ee5ee992a5c84cf5ece47db77a3856292b2a97d330dc9
2018-07-10 11:42:20,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 466345ab2be7bab12a61eac0a742c8f0ad1f18b722efe22ef22818016bed0b7913bb064134d228ec74b2535614c9b199
2018-07-10 11:42:21,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 6d72c993d97a883dd46c6bb85ee8cc8951133f3bd4c76027c29f44f28757624a0f1998bf0ad9111b4d7e6e7fe6bcb087
2018-07-10 11:42:22,423 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: cbc4296b53db8d66539fe1ef9dab5ba81d3b4f9629d1f2388cdbb31ef3baacd404d1d2a148f7d694ff83beec8750856e
2018-07-10 11:42:23,430 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 51d813ca09698322adfd9421c384cfe0b3e76763a683be4bdc1a767e853606e0bd9032bc2b2d04747f65a03708623357
2018-07-10 11:42:24,500 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 17e091912aef511aa3981a2010b53f51213c08481e3e72b248952c2518c0c6f1b72ac7c3de35091dd0d2f6d3c96c5d44
2018-07-10 11:42:25,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 3988fa955df11126878ac86ccade9c5099c9cdbe25cfbb487b76faf78c829dae4787ea1443700f58a637b8fa05b17408
2018-07-10 11:42:26,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 21b9748b63cebbb822e79429e948537952604889371e8799f93b743c4e64c59fcee366f38324bb488d1a9721e8c78967
2018-07-10 11:42:27,437 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 2c5a92ddc9a4c1e07266763e81466742c57f16f61e1cabc45b08b4cbea12f9a40ccd51ed97140869308f5a8724332329
2018-07-10 11:42:28,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: c16a3797685307d515bd8e934a481254b5e603371851837c5db13270aaf81e5980f9cd2be9ca79c6ef198c39e457bc0c
2018-07-10 11:42:29,430 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f8b1eb5af69bc89f018cfc1f67911cabe9cdfe1dd9c64b653584c90c1a5f31327fd2ccd1b8380abe80005616fae8d162
2018-07-10 11:42:30,510 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f2212f679cbb10899568455a21797ddf867f9a611d02c72955b42f6e27c5ce1bd030cb27e9a0a2f43e1147db575192d8
2018-07-10 11:42:31,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: dce5cae4647dbc322eb5266a332d94d2da7b7c9cbf0ab0c8605091294df240096fed2b4207f67bd48141d6345a83c798
2018-07-10 11:42:32,423 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: ffdf69571d26a1efdf9f4a33342f5e9bc5e7916da978f123a1e4ad9f598ba140ee7ca0095461250c8298aa91aaa0923d
2018-07-10 11:42:33,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 0cb4a9454a211816a22236045994c4b3b78996b44801eba418462704777ecabba19f7f878b9f65979bbd1ad12cdb6b00
2018-07-10 11:42:34,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: c6a1e4f2c07c596265f163ea534da5293e519eba0fac0c48cb063336fa8930c22e9baae4f39948ec4c138a362f70eafa
2018-07-10 11:42:35,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 4076dbc0c7766f2186614b20fc2af26fc16eae996d699751a56a1f42868295288644b4a45b26f54b777d2f2feeed2870
2018-07-10 11:42:36,431 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 7c3ab9565b40be44bc4f4ee1ed3ca991377c2318365d097535db5ba058d639cd75d499a2536b2a6c73c8b51fc4519cbc
2018-07-10 11:42:37,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e307abf7eee4a1d51ea793fcf8b9123c1c10872c8d21f5c69e83d194dc5236b4297d44da6cd649685e3e85f3a0605bab
2018-07-10 11:42:38,503 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 42cd925c3a0c6271defd2eeb4e8a5ecace030524de12ea952f7024157b233c083e7b801fb1a37525f605ff8175ecd749
2018-07-10 11:42:39,460 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: c14f4739c3908a3fd73ed66e1955232d377a8563f5d863f036b238cea0c5004e727b8757dd1a596207aae60cbf8b3a5d
2018-07-10 11:42:40,466 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 20f9077f7d4ca30e969911f8d3dd1341f9f7d5a988ccb9c5e387b31430325ad8eb13b044e978b37e5a6987728df926f2
2018-07-10 11:42:41,470 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e3abc16d77dc64d2a01b6d8bf2d3cc9bb15469a4384a662ba8342782c6e46a3a8d4c88f9234dbd340fbb53b728bfd0d4
2018-07-10 11:42:42,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: cbfdafb98e80b71947c5e850a7075bcce4db23c400328f083a4a3a6ef483f63ce7678629958151c3cebc45f920e92729
2018-07-10 11:42:43,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: fce7e20c7482c3f367bc18bb1a2fb6d8c25728e1b94315a9e847868fe54f1603027849ede24ced228aeaace329983639
2018-07-10 11:42:44,503 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 40c23211ea200fd0c6c2582aa70cf659c0b5ca60c7ae4f1f5405b4b385a34638e2147e7dbc666dcfd978c572788d3156
2018-07-10 11:42:45,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 613f5c326d75a13f5ee4e90a8fe56223ead60f91f722e882c949dd9e6781c03a4d43d3023b5990ce9bb5d682c71d7942
2018-07-10 11:42:46,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 3898b8a1863f855e63a57c6d703cb0589028864b4f8cd1f20a9ff9ef56143c946ab6e43bf26f167714e78dffba5f1856
2018-07-10 11:42:47,423 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 755b9d96baed36f5437cd3e71c538ce569d8be9c07561d61a0abbd79695fdabfc006faec770d85fe8f352e5baf84fef0
2018-07-10 11:42:48,469 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: fb4c99ef6cc215cba204a97f9c476909965008cab520a1ac5fd2a624aaff376e00fa8d903cc9b38448e1dd78b6b68a51
2018-07-10 11:42:49,477 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f54602b341f6eca40212c23ac4030d0859f2631abf731d276def53890c5fd90f78465bc307d5c88345a98db9fb7efd67
2018-07-10 11:42:50,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 65259c50cdea9fd8ce497b4aed270369e2791c22840457fb3015944d87e665077ad140cd6152224745b0aa950ec30e66
2018-07-10 11:42:51,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: d6e4b05612c23c6809f84d60bd967ce60985ed9c1336e7f9089029ac971d5704e1d08c6704a58e107d439598f3c8de90
2018-07-10 11:42:52,500 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: fbd88c35360d2f9b4d895825f3bcf71bd5343e577fd35b6a1479c2a712f6931164fb4c95ac7e78b16a400ca57a47be3a
2018-07-10 11:42:53,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b907b5b25848578b49644a8000a349c8b2210fcdb395f5494fbb22f650c76a29626f593cfeee189e01804ca0ffcfd96f
2018-07-10 11:42:54,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f17cd65d3673e62837a247ad8ae6b180223deb177b4c750b95f9aa7750a59577c6d1bb3cf48853193ce634e16345ed75
2018-07-10 11:42:55,423 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 14c78c242f2100f459db7b0b1a987851f5041ac85c75915cf7641e2056037dc3a6e62c1e0dd679f5ab430a00692c6538
2018-07-10 11:42:56,430 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: fec32712653d08245f340b80f49580614b87e414555598618c0d1c207cea369c6349caa695aa724ef916e404916fc343
2018-07-10 11:42:57,477 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 6bac33de07cd73d14593ec83df3f47d5fe698fbda9e6d2a5cc9ffea01dd097da77a88dbecbacd3a70f99c709d19ab3ea
2018-07-10 11:42:58,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 6cfd18c6fe4d959ce71b1843819c0857204b623307affe5184998f815b81a9ec208266bf4a779668b86818882bfeaf33
2018-07-10 11:42:59,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 595104d9cf2cf7581d1232bdf39f72ffdf2a37d8675b95cee998d01642a8b21909dfdbdf9182539942edc8388755e025
2018-07-10 11:43:00,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 559500803edf5b443813c110d89d0f4985e95dc2409f54fec41069009b5d342520c4ef0793ad9f14eb82a61f733b8c47
2018-07-10 11:43:01,505 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: eedc45bbdc575ebc5477124964a4c0f8ecd0914ce5359bbf12be671de59a1d8bc0d79bb31bdd0a320193aa0d7587759b
2018-07-10 11:43:02,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: a41fa613320a05a389d3a9efb6953fb5a0be1b1fd78964ed1b25b5f9b25887275de32898376879310b90b92bf816bd93
2018-07-10 11:43:03,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 9c5fa350d0c3f421547016464d638dd87bff2e13537edec806b99b6710309d67fc9354d7dfe1513f9c95583949480eb5
2018-07-10 11:43:04,460 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 564a5c4612aabb74e60d87d25904a4925e8d7f7a5d29fb5f8741a94b33b6455d8043d8b4c078cc31510086dc71e6c481
2018-07-10 11:43:05,467 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 09b3f8b3da2e09ae8ed6be1a81ba59c5c47e3c5de1221823918a3631c06e27c99787c5102600210cdea1c76d770f204b
2018-07-10 11:43:06,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e4fee3296aaa5d866342affd9638e414f4f35aa91ed06352d1efc7b980996f50fc3682c7533284175aa9ac3e244c41e6
2018-07-10 11:43:07,503 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b5ab0a5bb7e4bafc48e286420a6f1915406da273116924f5fbeb536e1baff5e1cb2607c5be6763bc5d685b34e425d1ef
2018-07-10 11:43:08,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 4ceacd89d98fa98878ded3c32c7cde911ec41f120e0d57704f35826b39afadde68e8b494a011f370c8c8df48c29ce18c
2018-07-10 11:43:09,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 42e485f031ad621df6fa3aded838e82e92315b0c50473134b5194c1b6791f38da4bee41845e2a59eda922fdc0d02bbc0
2018-07-10 11:43:10,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 3ee405cd549ecd7b5ebcb2b13b381b038c03ce27ed77c4dfeaf99849db1b53f11ff924b927324ff283df9986e37088d8
2018-07-10 11:43:11,431 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 21efdb979660dff164c715dcb575e42818327965dc97478051d0f2d15c131db9a7c2f655f35ac9a642759b0fc6190f90
2018-07-10 11:43:12,480 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 9260e3ce29b8904fb4914f685de5eb73b86c858a38f625b4e90f633cfd623faab48ed20b2d6a25ee42f992b46ab4276b
2018-07-10 11:43:13,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: fbfe672d0ab414f87fcd697f6abaa0f6107945034503a710dbb5f3917017b31f1450cc8a5d21af60740a301d2011f0bb
2018-07-10 11:43:14,433 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e609ea4cff3a234970221839360b8144bec5e5bc7938e681cc59bdce92a56c63a075a9c486c5e5d3f27f5b86f9255dd4
2018-07-10 11:43:15,503 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 2664d9c64c996533ed09f840bb794f6999c27928f317054b1f700d694990b3beded5bc292cd945f19d35a0db0b252493
2018-07-10 11:43:16,476 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 1a909be8f20b62529aa8b0b447c7a04af0c1dbcc7f9d427c8345b21d35bec5e95ac4792746d0a8fb9bcef90afa30ebd8
2018-07-10 11:43:17,470 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 97844ed611ef293665e50d008694a9161f3e6030097ca89534c01ee5c40ab37566ad1ce961453e4d765e9e9d6e96f821
2018-07-10 11:43:18,467 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 1f06ee8eb1d86b47bac79c962617f304499a9496eb97a049ce286d07fa7f98ac2a9065961d55055f9abab481e8509227
2018-07-10 11:43:19,470 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 1b9aa020998ea771eb82e32e251f4b60f3c3cda9416dbcd2f7ce659654014406d3836cdd51ccaac156cdd1e25121e36c
2018-07-10 11:43:20,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 44ce1d3d10398740a236d9bd54cb766e33230512368ba4abe36092d27e12ca895fda855d9a0a2e9b0ae1cc5df0ccba16
2018-07-10 11:43:21,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 6e78110028729d92a0043a36041bb2465c979e1cf4f31afa42ae17a56e8919bb77576f3f4b8c157fd36f070b7d435c25
2018-07-10 11:43:22,431 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e49deaac80599bc07419760606fafbbbb667f36c0cc79f3d2a300b969a5a88f87a3b42d064f9ae9e161ebd71fc2bde9a
2018-07-10 11:43:23,517 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 619f56313ad4d855fd47822fd46aa1e6172cdfe8ce46bb0513620c069e0a21a580e2510991e7ff8c8dcc07d1cc70609a
2018-07-10 11:43:24,480 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 61748b334fd7bfd10d37180dbfebf4da4631e274c6378287a1dde5f8d06368e9f032077c8ccb3ca14f494e158832670b
2018-07-10 11:43:25,470 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e60222f4108cbd98abfffb5482e41ee16c3122fa4b25a52d5e1b36e34efbf61d03ecd2981673161fa930f9164783ee53
2018-07-10 11:43:26,461 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 10cc8a0bb3e252504b7a0dcaacac376af17dd4cc67fea1297e1954fe857e18f8511ff062af3db4f5f856d0bec2d811fe
2018-07-10 11:43:27,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 0266dc0b8f8b404082f55aebc195b641efe80fcccdedc762d65d9bd27199490fc9f1b9139f7f7853935b077730bca994
2018-07-10 11:43:28,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 7da689ea1e13b8442e1763a9176a6de7202e62ce41ea2fea8d705d3d6f7fe281803df19045d7eb58edc9c3f8d59a7fec
2018-07-10 11:43:29,509 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b75647b01d65fa029fadf71bec744af4efe0ca5757d4134ec011026ae8ac860a2682e5f092d9aa18180514b7e1ab3315
2018-07-10 11:43:30,473 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 909a16192c136b96dbe696777d3aaecf0c57df32f72c4fc749aba16c1c93495e8862b37bfe8c41dd5e37e88008217e3a
2018-07-10 11:43:31,466 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 87fe3741304580d29aab9511c30dfde6830fd03a232389f9a57eb5967b7d8473161676fc198013ebd4f130da03faf06b
2018-07-10 11:43:32,467 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: d70671098e04a21c1e6005a6a0715db0c1c2a11f6d4b87fd678e7c9515535617545ffc84ad27a11237e727143f234f39
2018-07-10 11:43:33,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f6b960a90de3ecae16f29d9574b754711b2bd34c3ee11d2dec2eac5f1330ba2e0a3c7c86feed6ca3947f416955d2def2
2018-07-10 11:43:34,483 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 0d4c0226e7b529749bec6878b9324663897743a843ba9ad4df5a125cd50e399a46a5bf84748e9c8f6ce4868a6a7a79f3
2018-07-10 11:43:35,500 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: eb7daebaf6e7585842b9cba5f3564ac737c1e2138a18ef4e21df5ad77eeeffbe655e3df120e6ab1d58f90bfb04a8db1a
2018-07-10 11:43:36,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 7a0718cf26fc03a00b012099a7e68e75c50b78af4ad552bec74d0ba2e490235d9d3c329c38ee26ce9f3e44ecd3407850
2018-07-10 11:43:37,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: bc9da58d5e9659130269f7fd605091e08f635d5df062aca6fa1c40a8c93d4ec168eb97f5de7ac8ebf0899402d9a32478
2018-07-10 11:43:38,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 854e237baad226dec13c7264502bde6453d16f7ca5d03698ad770085d1e2c276d3fe4634d9f5b5a9582f913e63ff705e
2018-07-10 11:43:39,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: fec32cb08e2bdb9ddef5f312d2635672fee6db21c6ae4a171ed3422e89be487f96b6cefccc0caa00d5b32606811a9cbf
2018-07-10 11:43:40,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 848bb1d24e9e840eb6d94c1a83c756a3f96788b1343e26a26501063be577db7befce8ce4a018be89b353fce704400d84
2018-07-10 11:43:41,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e861b9e63a9b2dadcd84d69f258d1ed0c1165a5badad73918dd51b0a3bfaf8f095239cc8468ae68d49cdad0faf4897c7
2018-07-10 11:43:42,427 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 00197bb8f31cdd525e1b14dc8aa74a19985b1ebee4fe93fed79963b9d4d674251fa093accd071e703655e3de5a00d36e
2018-07-10 11:43:43,426 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 7073bd6adb1162191f1a2584de79e70ad82c823e236bbe6343c4639249f7a469178567becee0e2adec53f8ba70ce7509
2018-07-10 11:43:44,483 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 85455ba3c0d46f2d2a08ce9f933f942c79dc9d83f5594d916300487ece8868555ba435a139015c808e0addcb974b4a4f
2018-07-10 11:43:45,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: f17d4b3cb6a23cffde29904384943e47a9d311d0f008d0d5f34af932f2ee3dac1d71bbc1106ae87bbecc9956de5a6492
2018-07-10 11:43:46,493 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b52c0fe4f59cd3fa3c61cab568e710037fdd5f9b90f8800e56bcc6300a055c3ef165ddb858cd8e527c71aaafc166143f
2018-07-10 11:43:47,428 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b6d901fb9e85e4296d26e46a6bc6a4789bc609d17596ed598cf9adbdd20ea4a5391b366ab05257e70deb7aa7293d8d1c
2018-07-10 11:43:48,424 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: dc801e19cb06c7b097d4ad5cb4fa5e8c61f61a8d72f766ce3d8dcab05520ad8c21d20e7e91adfe5123a5fd58505a28d5
2018-07-10 11:43:49,507 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 43bbebcf313a06e1e4e4dcd5969fb12bc0566315922642776bda56d88a10530ae160cb06f3e9a370f67dbe71bbb6621a
2018-07-10 11:43:50,459 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 1d3985ed5c4026c16bd480f0fa316f7edc25e42705971f51f728598b43e6915572d42777a559f90f1b1a4a925fa83252
2018-07-10 11:43:51,463 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: bd7e3b9cfae1d93603467a7cfe2d8a37f56ff0498a36714c3519b34365bc8176f881ea7b15ca2126823ab64f62764ed1
2018-07-10 11:43:52,466 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: e66eb12ae146724988d96001d9a716dd266b4ee97186412a394e4d2d9e2ada6adc5daa86d403a9e1fa6f6117408fac8d
2018-07-10 11:43:53,470 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 647dce4b805269338dfc4df3365a4ba3e4c73adbd2576c3ae0ac701f1428c3824812e2f6955e3d781c9c1d0c95b042ca
2018-07-10 11:43:54,471 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 2913028fe8dde1db649c242a7c037bc4809b88e9225dd50ba10970236b9fb7dd86f6b758a255857959185e13d1b02020
2018-07-10 11:43:55,503 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: 740400f7f510c8f150102bf1ddc8e9ec4cebf89436a183ce5f417c2ae4888c947e9123172f497a4c3a7caf69a37c4bc0
2018-07-10 11:43:56,469 ERROR    lbrynet.core.HTTPBlobDownloader:64: [Mirror] Missing a blob: b44727569abfc24a43b39681b7b0536f5923c19d67a2294df6fdf151eed3967e5d3ba2a0cb9b07a9bc606e5de8447874
```

shyba commented:
rebased and updated logs to be less verbose while showing when a blob came from a mirror. Simple errors like a blob missing are now suppressed to debug and only success and hard errors are logged (when the blob is there but it fails downloading for real).

jackrobison commented on lbrynet/core/StreamDescriptor.py:
```diff
@@ -445,7 +445,10 @@ def download_sd_blob(session, blob_hash, payment_rate_manager, timeout=None):
+    mirror = HTTPBlobDownloader(session.blob_manager, [blob_hash], session.download_mirrors)
+    mirror.start()
     sd_blob = yield downloader.download()
+    mirror.stop()
```
What if the mirror finishes first? I think this should cancel whichever didn't finish. You could have the `start` and `download` calls in a deferred list set to fire on one callback

shyba replied:
It is done implicitly over blob manager. When one finishes a blob, the other will see that through the shared blob manager and go to the next (or halt, if nothing else to do). I can add that as a class documentation on the `HTTPBlobDownloader` class.

jackrobison commented on lbrynet/core/HTTPBlobDownloader.py:
```diff
@@ -0,0 +1,79 @@
+        self.blob_hashes = blob_hashes or []
+        self.looping_call = LoopingCall(self._download_next_blob_hash_for_file)
+        self.failures = 0
+        self.max_failures = 3
```
constants/settings?

jackrobison replied:
also, why more than one? if it fails, doesn't that mean the blob isn't mirrored (and thus won't succeed)?

jackrobison commented on lbrynet/core/HTTPBlobDownloader.py:
```diff
@@ -0,0 +1,79 @@
+            self.failures = 0
+        except Exception as exception:
+            self.failures += 1
+            log.error('Mirror failed downloading: %s', exception)
```
If you want the traceback you can do `log.exception("Mirror failed downloading")`
